### PR TITLE
Support the graceful stop for the indexnode

### PR DESCRIPTION
**Pull request by SimFG**
issue: #20550
Signed-off-by: SimFG <bang.fu@zilliz.com>

### CODE REVIEW

File: internal/indexnode/indexnode.go [hunk 1]
```diff
@@ -220,6 +223,18 @@ func (i *IndexNode) Start() error {
 
 // Stop closes the server.
 func (i *IndexNode) Stop() error {
+	if i.stopFlag.Load() {
```
Comment (jaime0815) [inline internal/indexnode/indexnode.go:226]: Should set sate code to StateCode_Stopping state before starting the stop method.

File: internal/indexcoord/node_manager.go [hunk 2]
```diff
@@ -34,16 +34,18 @@ import (
 
 // NodeManager is used by IndexCoord to manage the client of IndexNode.
 type NodeManager struct {
-	nodeClients map[UniqueID]types.IndexNode
-	pq          *PriorityQueue
-	lock        sync.RWMutex
-	ctx         context.Context
+	nodeClients   map[UniqueID]types.IndexNode
+	stoppingNodes map[UniqueID]struct{}
```
Comment (xiaofan-luan) [inline internal/indexcoord/node_manager.go:38]: when do you remove stopping nodes?

File: internal/indexcoord/node_manager.go [hunk 3]
```diff
@@ -73,11 +75,19 @@ func (nm *NodeManager) RemoveNode(nodeID UniqueID) {
...
 	nm.lock.Lock()
 	delete(nm.nodeClients, nodeID)
+	delete(nm.stoppingNodes, nodeID)
 	nm.lock.Unlock()
 	nm.pq.Remove(nodeID)
 	metrics.IndexCoordIndexNodeNum.WithLabelValues().Dec()
 }
 
+func (nm *NodeManager) StoppingNode(nodeID UniqueID) {
+	log.Debug("IndexCoord", zap.Any("Stopping node with ID", nodeID))
```
Comment (xiaofan-luan) [inline internal/indexcoord/node_manager.go:85]: also change RemoveNode to log.Info

File: internal/indexnode/indexnode.go [hunk 4]
```diff
@@ -220,22 +221,32 @@ func (i *IndexNode) Start() error {
...
 func (i *IndexNode) Stop() error {
-	// https://github.com/milvus-io/milvus/issues/12282
-	i.UpdateStateCode(commonpb.StateCode_Abnormal)
-	// cleanup all running tasks
-	deletedTasks := i.deleteAllTasks()
-	for _, task := range deletedTasks {
-		if task.cancel != nil {
-			task.cancel()
+	i.stopOnce.Do(func() {
+		i.UpdateStateCode(commonpb.StateCode_Stopping)
```
Comment (xiaofan-luan) [inline internal/indexnode/indexnode.go:225]: Add one log when going to stop

File: internal/indexcoord/node_manager.go (hunk 2 above)
Comment (SimFG) [reply internal/indexcoord/node_manager.go:38]: In the `RemoveNode` method, which is called when the node has stopped.
image

**Pull request merged**